### PR TITLE
Added functions to draw basic shapes

### PR DESCRIPTION
**Pull request by tpk32**
** closes issue** #6 

Added the following
-  `drawLine()`  : draws a line from one point to another
-  `drawRectangle()` :  draws a rectangle with given endpoints of a diagonal
-  `drawCircle()` : draws a circle with a given center and radius
-  `getPixelColor()` : returns the color of a pixel as MinGLColor variable
-  updated MinGLColor struct to compare colors using `==` and `!=` operators

### CODE REVIEW

File: MinGL.cpp [hunk 1]
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
 	glDisable(GL_SCISSOR_TEST);
 }
 
+const MinGLColor MinGL::getPixelColor(int x, int y, int width, int height) const{
+	int pixelSize = 4; // RGBA has 4 components (R, G, B, A)
```
Comment (amengol) [inline MinGL.cpp]: Please use constexpr whenever possible, like here

File: MinGL.cpp [hunk 2]
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
...
+    unsigned char* pixel = new unsigned char[pixelSize];
```
Comment (amengol) [inline MinGL.cpp]: This is clearly leaking. Clean up or better yet, use a vector.

File: MinGL.cpp [hunk 3]
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
 	glDisable(GL_SCISSOR_TEST);
 }
 
+const MinGLColor MinGL::getPixelColor(int x, int y, int width, int height) const{
```
Comment (amengol) [inline MinGL.cpp]: Not used in this function. Please remove it.

File: MinGL.h [hunk 4]
```diff
@@ -1,10 +1,25 @@
...
+			if( fabs(this->rgba[i] - color2.rgba[i]) > epsilon )return false;
+		}
+		return true;
+	}
+	bool operator != (const MinGLColor color2) const{
+		for(int i=0; i<4; i++){
+			if( fabs(this->rgba[i] - color2.rgba[i]) > epsilon ) return true;
+		}
+		return false;
+	}
```
Comment (amengol) [inline MinGL.h]: Simplify to 
`return !(*this == color2);`

File: MinGL.cpp [hunk 5]
```diff
@@ -169,6 +170,103 @@ void MinGL::putPixel(int x, int y, const MinGLColor& color, int width, int heigh
...
+		putPixel(xa, i, color);
+	}
+}
+
+
+void MinGL::drawCircle(int xc, int yc, int rad, const MinGLColor& color){
+	// specify center (x,y) of circle and its radius
+	// Bresenham's circle drawing algorithm is used here
+
+	auto draw = [&](int x, int y){
```
Comment (amengol) [inline MinGL.cpp]: No need to capture everything by reference here, since most are just integers. If you must, capture only color by reference.

**Review by amengol — COMMENTED** [review]
Thanks for contributing! :)

**Review by amengol — CHANGES_REQUESTED** [review]
Please address review comments.

File: MinGL.cpp (hunk 3 above)
Comment (tpk32) [reply MinGL.cpp]: thanks for the review, I've removed `width` and `height` parameters, also removed the redundant `#include<queue>`

File: MinGL.cpp (hunk 1 above)
Comment (tpk32) [reply MinGL.cpp]: used constexpr for `pixelsize`  and static constexpr for `epsilon` in MinGLColor struct

File: MinGL.cpp (hunk 2 above)
Comment (tpk32) [reply MinGL.cpp]: changed to vector

File: MinGL.cpp (hunk 5 above)
Comment (tpk32) [reply MinGL.cpp]: captured only color by reference

File: MinGL.h (hunk 4 above)
Comment (tpk32) [reply MinGL.h]: done

**Comment by tpk32** [comment]
Changes made:
- `epsilon` in MinGLColor struct is declared as `static constexpr`
- `getPixelColor()`  returns `MinGLColor` instead of `const MinGLColor` 
- `drawLine()`, `drawRectangle()` and `drawCircle()` functions are now declared `const` 
- `pixelSize` in `getPixelColor()` is declared `constexpr`
- used `vector`  in `getPixelColor()` instead of allocating memory for char array
- `dx` and `dy` in `drawLine()` are now const
- `xa`, `ya`, `xb`, `yb` in `drawRectangle()` are now const
- `draw` lambda in `drawCircle()` captures `color` by reference and `xc`,  `yc`, and `this` pointer by value

**Comment by tpk32** [comment]
Please review the changes and let me know if any other adjustments are required.

File: Example/shapesExample.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+int main()
+{
+	// Window sizes
+	const int width = 800;
+	const int height = 600;
+	MinGL minGL;
+	if (!minGL.init(width, height, "MinGL"))
+		return -1;
+
+	const char* glsl_version = "#version 130";
```
Comment (amengol) [inline Example/shapesExample.cpp]: Not sure why this glsl_version line is here!

**Review by amengol — CHANGES_REQUESTED** [review]

**Review by amengol — CHANGES_REQUESTED** [review]
Looking good. Just address the glsl version comment. Thanks!

File: Example/shapesExample.cpp (hunk 6 above)
Comment (tpk32) [reply Example/shapesExample.cpp]: this line was included in [Main.cpp](https://github.com/tpk32/MinGL/blob/main/Example/Main.cpp) and i copied some code from there, that's why its here also. I don't know if it serves some purpose but the code runs fine without it.
Shall i remove this line from both [Main.cpp](https://github.com/tpk32/MinGL/blob/feature-MinGL-adding-basic-shapes/Example/Main.cpp) and [shapesExample.cpp](https://github.com/tpk32/MinGL/blob/feature-MinGL-adding-basic-shapes/Example/shapesExample.cpp)

Comment (amengol) [reply Example/shapesExample.cpp]: Hmmm. I see now. Yeah, it was probably a leftover from something I was trying to do at the time.
Can you please remove from both?

Now that I see it, I don't like that it has two main() functions. We're not JAVA! :)
I might be asking too much, but can you try moving the examples to a Header file with functions like drawCube and DrawShapes and call them from "Main.cpp"? You can leave one of the examples commented out, no problem.
Thanks a lot!

**Comment by tpk32** [comment]
Changes made:
- created `Example.h` and `Example.cpp` for example functions
- changed directory structure to keep things clean
  - `MinGL.h` and `Example.h` header files moved to `include` folder
  - `MinGL.cpp` and `Example.cpp`  files moved to `src` folder
- changed `Main.cpp` to use example functions from `Example.h` header
- updated `readme.md`
  - changed look of directory structure
  - added compiling instructions

**Review by amengol — CHANGES_REQUESTED** [review]
I appreciate what you did, really I do, but this is a library and you restructured it like a runtime C++ project.
In other words, I think MinGL.h and MinGL.cpp (and other core files) are better off in the main level as before. However, I am fine with an "Example" folder with everything example related in it. Sound good?

**Comment by tpk32** [comment]
Sure, I'll make the changes today. I am really loving working on MinGL. Also helps me learn a lot from your experience. Haha..

**Comment by tpk32** [comment]
Changes made:
- Moved `MinGL.cpp` and `MinGL.h` to main level as before
- Moved `Example.cpp` and `Example.h` to Example folder
- Updated compiling instructions in readme.md

File: README.md [hunk 7]
```diff
@@ -9,27 +9,62 @@ Note: Since ray-tracing is very demanding, you better of writing an image to dis
...
+├── glad
+│   ├── glad.c
+│   ├── glad.h
+│   └── khrplatform.h
+├── GLFW
+│   ├── glfw3.h
+│   └── glfw3native.h
+├── MinGL.h
+├── MinGL.cpp
+└── Main.cpp
```
Comment (amengol) [inline README.md]: Main.cpp is in the Example folder. Please correct this README to reflect it.

**Review by amengol — CHANGES_REQUESTED** [review]
This is almost perfect. Just a README mistake and that's it. Thank you so much.

File: README.md (hunk 7 above)
Comment (tpk32) [reply README.md]: Done, please tell me if any other change is required.

**Review by amengol — APPROVED** [review]
That looks very good. Thank you so much!

**Pull request merged**